### ISSUE TITLE
tf.gradients returns None in tf.map_fn

### ISSUE THREAD

**Issue by cjf00000** [tensorflow/tensorflow#3972]
GitHub issues are for bugs / installation problems / feature requests.  
For general support from the community, see [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow).
To make bugs and feature requests more easy to find and organize, we close issues that are deemed
out of scope for GitHub Issues and point people to StackOverflow.

For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.
### Environment info

Operating System:
Mac OS 10.10

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
None

If installed from binary pip package, provide:
1. Which pip package you installed.
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
   0.10.0rc0

If installed from source, provide 
1. The commit hash (`git rev-parse HEAD`)
2. The output of `bazel version`
### Steps to reproduce

I want to evaluate the diagonal of Hessian with `tf.map_fn`, which takes a function that maps each dimension (scalar) to its Hessian

```
import tensorflow as tf


def hessian_factory(f):
    def hessian1(x):
        g = tf.gradients(f, x)[0]
        h = tf.gradients(g, x)[0]
        return h

    return hessian1

sess = tf.Session()
x = tf.Variable(1.0)
sess.run(tf.initialize_all_variables())

f = 1. / (1 + tf.exp(-x))
func = hessian_factory(f)
h = tf.map_fn(func, x)

print(sess.run(h))
```

However, it seems `tf.gradients(f, x)` produces `None`.

Even the following statements gives `None`

```
import tensorflow as tf

sess = tf.Session()
x = tf.Variable(1.0)
sess.run(tf.initialize_all_variables())

f = 1. / (1 + tf.exp(-x))
h = tf.map_fn(lambda x: tf.gradients(f, x)[0], x)
```
### What have you tried?

1.
### Logs or other output that would be helpful

(If logs are large, please upload as attachment).

**Comment by yuanbyu** [comment]
It is a known issue that tf.gradients() can't be called inside map_fn, due to some tricky problems with TensorArray.  However you can call tf.gradients() inside tf.while_loop so try to use tf.while_loop directly.

**Comment by drpngx** [comment]
Is that still a problem with recent versions?

**Comment by yaroslavvb** [comment]
`None`s in Hessians are expected for reasons unrelated to `map_fn` -- for implementation reasons, TensorFlow returns `None` instead of `0` when the gradient is not a function of the variable being differentiated against. Hence off-diagonal terms are None's, see discussion here for Hessians -- https://www.google.com/url?hl=en&q=https://groups.google.com/a/tensorflow.org/d/msgid/discuss/d83404b7-793a-4201-ad73-fcecbb83eaac%2540tensorflow.org?utm_medium%3Demail%26utm_source%3Dfooter&source=gmail&ust=1485411137547000&usg=AFQjCNEhO37F5hU6GDwsfv8BBbAariqCFQ

Also see https://github.com/tensorflow/tensorflow/issues/783 for why it's None's instead of 0's

**Comment by yaroslavvb** [comment]
You need to do something like this to get Hessians

```
def replace_none_with_zero(l):
  return [0 if i==None else i for i in l] 

tf.reset_default_graph()

x = tf.Variable(1.)
y = tf.Variable(1.)
loss = tf.square(x) + tf.square(y)
sess = create_session()
grads = tf.gradients([loss], [x, y])
hess0 = replace_none_with_zero(tf.gradients([grads[0]], [x, y]))
hess1 = replace_none_with_zero(tf.gradients([grads[1]], [x, y]))
hessian = tf.pack([tf.pack(hess0), tf.pack(hess1)])
print hessian.eval()

```

**Comment by drpngx** [comment]
Closing since @yaroslavvb comment seems like a reasonable answer.

**Comment by RylanSchaeffer** [comment]
@aselle and @drpngx , can something about the interaction between tf.map_fn and gradients be added to the API documentation? I don't think anything is currently mentioned.

**Comment by drpngx** [comment]
So there was a convention where gradients would return none (and not zero)
if integers were involved. Then we changed some case for correctness. That
happened maybe a couple of months ago. Do your computations involve
integers?

**Comment by RylanSchaeffer** [comment]
@drpngx , I'm actually dealing with a different use case (I was using `tf.map_fn` to feed a sequence of 3d images through a few `conv2d` layers and just discovered that no gradients were reaching my convolutional layers). Was your question intended for me?

Also, as an aside, how is a sequence of images normally fed through convolutional layers?

**Comment by RylanSchaeffer** [comment]
I created a separate issue for my question/suggestion. https://github.com/tensorflow/tensorflow/issues/19896

**Comment by drpngx** [comment]
I see. I don't really see why `map_fn` can't propagate the gradients. Maybe file a feature request and CC @skye.

**Comment by RylanSchaeffer** [comment]
Sounds good! I opened a new feature request.
